commands/build.go

```diff
@@ -166,26 +167,47 @@ func runBuild(cmd *cobra.Command, args []string) error {
 	return nil
 }
 
-func build(services *stack.Services, queueDepth int, shrinkwrap bool) {
+func build(services *stack.Services, queueDepth int, shrinkwrap bool) error {
 	wg := sync.WaitGroup{}
 
-	workChannel := make(chan stack.Function)
+	type indexedFunction struct {
+		Index    int
+		Function stack.Function
+	}
+
+	workChannel := make(chan indexedFunction)
+
+	type namedError struct {
+		FunctionName string
+		Err          error
+		Skipped      bool
+	}
+
+	errorList := make([]namedError, len(services.Functions))
+	hasError := false
 
 	wg.Add(queueDepth)
 	for i := 0; i < queueDepth; i++ {
 		go func(index int) {
-			for function := range workChannel {
+			for indexedFunc := range workChannel {
+				function := indexedFunc.Function
 				fmt.Printf(aec.YellowF.Apply("[%d] > Building %s.\n"), index, function.Name)
+				var err error
 				if len(function.Language) == 0 {
-					fmt.Println("Please provide a valid language for your function.")
+					err = fmt.Errorf("Please provide a valid language for your function")
+					log.Println(err)
 				} else {
 
 					combinedBuildOptions := combineBuildOpts(function.BuildOptions, buildOptions)
-					err := builder.BuildImage(function.Image, function.Handler, function.Name, function.Language, nocache, squash, shrinkwrap, buildArgMap, combinedBuildOptions, tag)
+					err = builder.BuildImage(function.Image, function.Handler, function.Name, function.Language, nocache, squash, shrinkwrap, buildArgMap, combinedBuildOptions, tag)
 					if err != nil {
 						log.Println(err)
 					}
 				}
+				errorList[indexedFunc.Index] = namedError{FunctionName: function.Name, Err: err}
+				if err != nil {
+					hasError = true
```

Comment:
Does it work as expected without lock?
This might have random behaviour.

Reply:
Concerning `errorList`, the index unique to each function build, so they do not step over each other. Here is a simplifying example:

```go
package main

import (
	"fmt"
	"sync"
)

func main() {
	queueDepth := 2
	numInts := 20

	wg := sync.WaitGroup{}
	intList := make([]int, numInts)
	intChannel := make(chan int)

	wg.Add(queueDepth)
	for i := 0; i < queueDepth; i++ {
		go func(index int) {
			for v := range intChannel {
				intList[v] = v
			}
			wg.Done()
		}(i)
	}

	for j := 0; j < numInts; j++ {
		intChannel <- j
	}
	close(intChannel)

	wg.Wait()

	fmt.Println(intList)
}

```

Concerning `hasError`, the current implementation does not need a lock since `hasError` acts as a flag that is not concern with the order it is set. But to make this easier for future developers, I refactored the code to have `hasError` defined after all the work is done.